packages/cli/src/util/dev/server.ts

```diff
@@ -1536,9 +1537,34 @@ export default class DevServer {
 
             // Retain orginal pathname, but override query parameters from the rewrite
             const beforeRewriteUrl = req.url || '/';
-            const rewriteUrlParsed = url.parse(beforeRewriteUrl);
-            rewriteUrlParsed.search = url.parse(rewritePath).search;
-            req.url = url.format(rewriteUrlParsed);
+
+            if (isURL(rewritePath)) {
+              const rewriteUrlParsed = new URL(rewritePath);
+              const rewriteOrigin = `${rewriteUrlParsed.protocol}//${rewriteUrlParsed.host}`;
+
+              const originalUrlParsed = new URL(
+                `http://127.0.0.1/${beforeRewriteUrl}`
+              );
+              const isSameOrigin =
+                `${originalUrlParsed.protocol}//${originalUrlParsed.host}` ===
+                rewriteOrigin;
+
+              if (isSameOrigin) {
+                req.url = rewritePath.replace(rewriteOrigin, '');
+              } else {
+                // Proxy to absolute URL with different origin
+                debug(`ProxyPass: ${rewritePath}`);
+                this.setResponseHeaders(res, requestId);
+                proxyPass(req, res, rewritePath, this, requestId);
+                return;
+              }
+            } else {
+              // Retain orginal pathname, but override query parameters from the rewrite
+              const rewriteUrlParsed = url.parse(beforeRewriteUrl);
+              rewriteUrlParsed.search = url.parse(rewritePath).search;
+              req.url = url.format(rewriteUrlParsed);
```

Comment:
We're keeping the broken path writes in place because it matches production. This will be fixed in production, then in `vc dev`, here: https://github.com/vercel/vercel/pull/8456